docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1056 @@
+    + [Access violation testing](#access-violation-av-testing)
+  * [Loading and storing vectors](#loading-and-storing-vectors)
+    + [Loading](#loading)
+    + [Storing](#storing)
+    + [Casting](#casting)
+  * [Mindset](#mindset)
+    + [Edge cases](#edge-cases)
+    + [Scalar solution](#scalar-solution)
+    + [Vectorized solution](#vectorized-solution)
+  * [Toolchain](#toolchain)
+    + [Creation](#creation)
+    + [Bit operations](#bit-operations)
+    + [Equality](#equality)
+    + [Comparison](#comparison)
+    + [Math](#math)
+    + [Conversion](#conversion)
+    + [Widening and Narrowing](#widening-and-narrowing)
+    + [Shuffle](#shuffle)
+  * [Summary](#summary)
+    + [Best practices](#best-practices)
+
+TL;DR: Go to [Summary](#summary)
+
+# Introduction to vectorization with Vector128 and Vector256
+
+Vectorization is the art of converting an algorithm from operating on a single value at a time to operating on a set of values (vector). It can greatly improve performance at a cost of increased code complexity.
+
+In recent releases, .NET has introduced many new APIs for vectorization. The vast majority of them are hardware specific, so they require users to provide an implementation per processor architecture (x64 and/or arm64), with the option of using the most optimal instructions for hardware that is executing the code.
+
+.NET 7 introduced a set of new APIs for `Vector128<T>` and `Vector256<T>` for writing hardware-agnostic, cross platform vectorized code. The purpose of this document is to introduce you to the new APIs and provide a set of best practices.
+
+## Code structure
+
+`Vector128<T>` represents a 128-bit vector of type `T`. `T` is constrained to specific primitive types:
+
+* `byte` and `sbyte` (8 bits).
+* `short` and `ushort` (16 bits).
+* `int`, `uint` and `float` (32 bits).
+* `long`, `ulong` and `double` (64 bits).
+* `nint` and `unit` (32 or 64 bits, depending on the architecture)
+
+A single `Vector128` operation allows you to operate on: 16 (s)bytes, 8 (u)shorts, 4 (u)ints/floats, or 2 (u)longs/double(s).
+
+```
+------------------------------128-bits---------------------------
+|             64                |               64              |
+-----------------------------------------------------------------
+|      32       |      32       |      32       |      32       |
+----------------------------------------------------------------|
+|  16   |  16   |  16   |  16   |  16   |  16   |  16   |  16   |
+-----------------------------------------------------------------
+| 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 |
+-----------------------------------------------------------------
+```
+
+`Vector256<T>` is twice as big as `Vector128<T>`, so when it is hardware accelerated, and the data is large enough, you should use it instead of `Vector128<T>`. To check the acceleration, use `Vector128.IsHardwareAccelerated` and `Vector256.IsHardwareAccelerated` properties.
```

Comment:
This can be tricky advice. It being larger isn't necessarily always a win due to various "quirks" that exist in the underlying implementation.

Namely, `Vector256<T>` on x86/x64 is mostly treated as `2x Vector128<T>` and while there are some operations that can "cross lanes", they can sometimes be more expensive or have other hidden costs.

So while it can often be a win to switch when the data is large enough, there are some algorithms where that "breaks down" and it can end up being much slower.